experiment/build/generate_cloudbuild.py

```diff
@@ -13,65 +13,126 @@
+            # Base images are built before creating fuzzer benchmark builds,
+            # so it's not required to wait for them to build.
+            if 'base' in dep and not build_base_images:
+                continue
+            step['wait_for'] += [dep]
+
         cloud_build_spec['steps'].append(step)
+        cloud_build_spec['images'].append(
+            get_experiment_tag_for_image(image_specs))
+        cloud_build_spec['images'].append(
+            get_experiment_tag_for_image(image_specs).split(':')[0])
+
+    if any('coverage' in image_name for image_name in image_templates.keys()):
+        cloud_build_spec['steps'] += coverage_steps(benchmark)
 
     return cloud_build_spec
 
 
-def main():
-    """Generates Cloud Build specification."""
-    parser = argparse.ArgumentParser(description='GCB spec generator.')
-    parser.add_argument('-r',
-                        '--docker-registry',
-                        default='gcr.io/fuzzbench/',
-                        help='Docker registry to use.')
-    args = parser.parse_args()
-
-    # TODO(Tanq16): Create fuzzer/benchmark list dynamically.
-    fuzzers = ['afl', 'libfuzzer']
-    benchmarks = ['libxml', 'libpng']
-    buildable_images = docker_images.get_images_to_build(fuzzers, benchmarks)
-    cloud_build_spec = create_cloud_build_spec(buildable_images,
-                                               args.docker_registry)
-    # Build spec can be yaml or json, use whichever:
-    # https://cloud.google.com/cloud-build/docs/configuring-builds/create-basic-configuration
-    print(yaml.dump(cloud_build_spec))
-    print(json.dumps(cloud_build_spec))
-
-
-if __name__ == '__main__':
-    main()
+def _get_buildable_images(fuzzers, benchmarks):
+    return docker_images.get_images_to_build(fuzzers, benchmarks)
+
+
+def generate_base_images_build_spec():
+    """Returns build spec for base images."""
+    buildable_images = _get_buildable_images([''], [''])
+    image_templates = {'base-image': buildable_images['base-image']}
+    return create_cloud_build_spec(image_templates, build_base_images=True)
+
+
+def generate_benchmark_build_spec(fuzzers, benchmarks):
+    """Returns build spec for fuzzer-benchmark builds."""
+    buildable_images = _get_buildable_images(fuzzers, benchmarks)
+    image_templates = {}
+    for name in buildable_images:
+        if any(image_type in name for image_type in ('base',
+                                                     'coverage', 'dispatcher')):
+            continue
+        image_templates[name] = buildable_images[name]
+    return create_cloud_build_spec(image_templates)
+
+
+def generate_coverage_build_spec(fuzzers, benchmarks):
+    """Returns build spec for coverage builds."""
+    buildable_images = _get_buildable_images(fuzzers, benchmarks)
+    image_templates = {
+        name: image for name, image in buildable_images.items() if 'coverage' in name
+    }
+    return create_cloud_build_spec(image_templates, benchmark=benchmarks[0])
```

Comment:
why do we need separate specs for base images, benchmark images, coverage images, etc? can we generate a single spec that we pass to `gcloud build submit`?

Reply:
According to what is being built, the `gcb_build.py` calls one of these functions (generate base images, benchmarks or coverage). That spec is written to a file and passed to `gcloud build submit`. This is because `gcb_build.py` has similar functions. 
By any chance do you mean I should move the code from these functions to those in `gcb_build.py` ?

Reply:
Yes. As discussed earlier, would it be possible to replace the whole gcb_build.py script with something like:
```
buildable_images = docker_images.get_images_to_build(fuzzers, benchmarks)	
build_spec = create_cloud_build_spec(buildable_images, args.docker_registry)
temp_file = create_temp_file(`build_spec.yaml`)
yaml.dump(build_spec, temp_file)
subprocess.call(gcloud build submit `build_spec.yaml`)
```
?

Reply:
I don't think so, that would try building everything on one instance which defeats the purpose of GCB.

Reply:
+1. But I think the closest possible to this would be to use the three functions from gcb_build.py itself to do the work of the three functions here. I'll try to get that working today.

Reply:
Can you clarify what you mean by that? My understanding is that by setting the image dependencies (with `waitFor`-s) we'd be building images concurrently (up to 30 concurrent builds -- which should be our current quota).
https://cloud.google.com/cloud-build/docs/configuring-builds/configure-build-step-order
What am i missing?

Reply:
IIUC, when we build the fuzzer benchmarks, we don't keep base images as dependencies i.e., they are not in wait for because they are assumed to be already built previously. So when passing the complete build spec, that would also build the base images along with coverage builds (even when not required).

Reply:
> So when passing the complete build spec, that would also build the base images along with coverage builds (even when not required).

Buildkit's caching mechanism would avoid rebuilding redundant images/layers, no?

> IIUC, when we build the fuzzer benchmarks, we don't keep base images as dependencies i.e., they are not in wait for because they are assumed to be already built previously. 

I'm not sure it's a good idea to remove/hide dependencies like that. What if we do change the base image? We'd need to always remember to manually rebuild/push it?

Reply:
I think the wait_for stuff waits for things that happen on the same machine. Do you think otherwise.

Reply:
> I think the wait_for stuff waits for things that happen on the same machine. Do you think otherwise?

yes. GCB abstracts away the actual machines the build steps are executed on, no?